Adding Terraform Plugin Protocol top-level section (#241) (#242)
hashicorp/terraform-docs-common: website/docs/plugin/terraform-plugin-protocol.mdx

| `Configure`                  | `ConfigureProvider`          | Passes the practitioner-specified provider configuration to the provider.                                                                                                                                                            |
| `ReadResource`               | `ReadResource`               | Called when refreshing a resource's state.                                                                                                                                                                                           |
| `ReadDataSource`             | `ReadDataSource`             | Called when refreshing a data source's state.                                                                                                                                                                                        |
| `PlanResourceChange`         | `PlanResourceChange`         | Calculates a plan for a resource. A proposed new state is generated, which the provider can modify.                                                                                                                                  |

> | `ApplyResourceChange`        | `ApplyResourceChange`        | Called when a diff between the resource's state and configuration has been detected, and the practitioner has approved a planned change. The provider is to apply the changes contained in the plan, and return the resulting state. |

Reviewer: Terraform allows for plans without configuration changes (e.g. only updating `Computed`-only values). This is so providers can signal potential provider behavior changes via a plan-apply workflow. I believe the below wording should capture what we are going for here.

```suggestion
| `ApplyResourceChange`        | `ApplyResourceChange`        | Called when a practitioner has approved a planned change. The provider is to apply the changes contained in the plan, and return a resulting state matching the given plan. |
```

| `ImportResourceState`        | `ImportResourceState`        | Called when importing a resource into state so that the resource becomes managed.                                                                                                                                                    |

![diagram: Terraform Plugin Protocol 6 RPCs](/img/docs/terraform-plugin-protocol-6-rpcs-all.png)

## RPCs and Terraform commands

The following illustrate the RPCs that are sent during execution of the specified Terraform commands.

The RPCs issued are for a provider that is using version 6 of the Terraform plugin protocol when the Terraform configuration contains configuration for the provider, a resource and a data source.

### _terraform validate_

Executing _terraform validate_ results in the following unique RPCs:

![diagram: Terraform Plugin Protocol 6 Validate RPCs](/img/docs/terraform-plugin-protocol-6-rpcs-validate.png)

The general sequence of RPCs is as follows:

- GetProviderSchema
- ValidateProviderConfig
- ValidateResourceConfig
- ValidateDataResourceConfig

### _terraform plan_

Executing _terraform plan_ results in the following unique RPCs:

> ![diagram: Terraform Plugin Protocol 6 Plan RPCs](/img/docs/terraform-plugin-protocol-6-rpcs-plan.png)

Reviewer: This diagram is missing `ReadResource`

Author: Well spotted! Fixed.

The general sequence of RPCs is as follows:

- GetProviderSchema
- ValidateProviderConfig
- ValidateResourceConfig
- ValidateDataResourceConfig
- ConfigureProvider
- ReadDataSource
- PlanResourceChange

### _terraform apply_

Executing _terraform apply_ (with default -refresh=true and -auto-approve) results in the following unique RPCs:

![diagram: Terraform Plugin Protocol 6 Plan RPCs](/img/docs/terraform-plugin-protocol-6-rpcs-apply.png)

The general sequence of RPCs is as follows:

- GetProviderSchema
- ValidateProviderConfig
- ValidateResourceConfig
- ValidateDataResourceConfig
- ConfigureProvider
- UpgradeResourceState
  - only called if resource already exists in state
- ReadResource

>   - only called if resource already exists in state)

Reviewer: ```suggestion
  - only called if resource already exists in state
```

- ReadDataSource
- PlanResourceChange
- ApplyResourceChange

> [0]: https://en.wikipedia.org/wiki/Static_build#Static_building
> 
> [1]: https://golang.org/

Reviewer: These detached links are not present on this new page.

```suggestion
```

* [tf6muxserver](/plugin/mux/combining-protocol-version-6-providers): A package to combine multiple protocol version 6 providers.

## Protocol Version 5

Protocol version 5 is compatible with Terraform CLI version 0.12 and later.

Implementations include:

* [terraform-plugin-framework](/plugin/framework): A higher-level SDK that makes Terraform provider development easier by abstracting implementation details.
* [terraform-plugin-sdk/v2](/plugin/sdkv2): A higher-level SDK that makes Terraform provider development easier by abstracting implementation details.
* [terraform-plugin-go tf5server](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server): A lower-level SDK to develop Terraform providers for more advanced use cases.
* [tf6to5server](/plugin/mux/translating-protocol-version-6-to-5): A package to translate protocol version 6 providers into protocol version 5.
* [tf5muxserver](/plugin/mux/combining-protocol-version-5-providers): A package to combine multiple protocol version 5 providers.

## Protocol RPCs

The Terraform plugin protocol specifies the following RPCs:

> | Protocol Version 5           | Protocol Version 6           |

Reviewer: It might be helpful if this table gave a brief description of the RPC in a new column. 👍 e.g. `GetProviderSchema` Returns schema information for the entire provider, including the provider schema, provider meta schema, all data source schemas, and all resource schemas.

| `ValidateDataSourceConfig`   | `ValidateDataResourceConfig` |
| `UpgradeResourceState`       | `UpgradeResourceState`       |
| `Configure`                  | `ConfigureProvider`          |
| `ReadResource`               | `ReadResource`               |
| `ReadDataSource`             | `ReadDataSource`             |
| `PlanResourceChange`         | `PlanResourceChange`         |
| `ApplyResourceChange`        | `ApplyResourceChange`        |
| `ImportResourceState`        | `ImportResourceState`        |

![diagram: Terraform Plugin Protocol 6 RPCs](/img/docs/terraform-plugin-protocol-6-rpcs-all.png)

## RPCs and Terraform commands

The following illustrate the RPCs that are sent during execution of the specified Terraform commands.

The RPCs issued are for a provider that is using version 6 of the Terraform plugin protocol when the Terraform configuration contains configuration for the provider, a resource and a data source.

### _terraform validate_

Executing _terraform validate_ results in the following unique RPCs:

![diagram: Terraform Plugin Protocol 6 Validate RPCs](/img/docs/terraform-plugin-protocol-6-rpcs-validate.png)

> The actual sequence of RPCs is as follows:

Reviewer: Terraform implementation details will dictate the actual order and are not protected by any sort of contracts with providers. The ordering is determined by the DAG and any parallelism. In fact, we may be want to be even more generic and just mention that here's the _potential_ RPCs. If there are no data sources in the configuration, then `ValidateDataResourceConfig` is not called, for example.

```suggestion
The general sequence of RPCs is as follows:
```

- [GetProviderSchema](#getproviderschema-rpc)

> - [GetProviderSchema](#getproviderschema-rpc)

Reviewer: The duplicate calls to the RPC are an implementation detail of how Terraform currently works and not part of any contract with providers. I'd suggest removing duplicates here and below.

- [ValidateProviderConfig](#validateconfig-rpcs)
- [ValidateResourceConfig](#validateconfig-rpcs)
- [ValidateDataResourceConfig](#validateconfig-rpcs)

> The correlation between the RPCs and the Terraform plugin framework methods is as follows:

Reviewer: I believe we should leave framework specific information onto its own framework documentation, as this top-level documentation section is serving terraform-plugin-go, terraform-plugin-sdk, terraform-plugin-framework, and other generic gRPC implementations.

### _terraform apply_

Executing _terraform apply_ (with default -refresh=true and -auto-approve) results in the following unique RPCs:

![diagram: Terraform Plugin Protocol 6 Plan RPCs](/img/docs/terraform-plugin-protocol-6-rpcs-apply.png)

The actual sequence of RPCs is as follows:

- [GetProviderSchema](#getproviderschema-rpc)
- [GetProviderSchema](#getproviderschema-rpc)
- [ValidateProviderConfig](#validateconfig-rpcs)
- [ValidateResourceConfig](#validateconfig-rpcs)
- [ValidateDataResourceConfig](#validateconfig-rpcs)
- [GetProviderSchema](#getproviderschema-rpc)
- [ValidateProviderConfig](#validateconfig-rpcs)
- [ConfigureProvider](#configureprovider-rpc)
- UpgradeResourceState
  - only called if resource already exists in state
- [ValidateResourceConfig](#validateconfig-rpcs)
- [ValidateDataResourceConfig](#validateconfig-rpcs)
- [ReadResource](#read-rpcs)
  - only called if resource already exists in state)
- [ReadDataSource](#read-rpcs)
- [PlanResourceChange](#planresourcechange-rpc)

>   - if _-auto-approve_ is not used this is the point at which Terraform prompts the practitioner to confirm the proposed changes

Reviewer: We may want to confer with the core team to determine if they want to document this level of detail. For provider developers, I believe the important details are which RPCs are potentially called.

Author: I'll remove this for now.

| `ConfigureProvider`                                                               | `provider.Provider` interface `Configure` method                                                                                                                   |
| `ReadResource` / `ReadDataSource`                                                 | `resource.Resource` and `datasource.DataSource` interface `Read` method                                                                                            |
| `PlanResourceChange`                                                              | `resource.Resource` interface `Schema` method plan modifiers and `ModifyPlan` method                                                                               |
| `ApplyResourceChange`                                                             | `resource.Resource` interface `Create`, `Update`, or `Delete` method                                                                                               |

> ## RPCs and Terraform plugin framework

Reviewer: This seems like it should be its own framework documentation page since its SDK specific and could change over time.
